doc/optimization.md

```diff
@@ -0,0 +1,86 @@
+# RaptorJIT optimization guide
+
+The first step to optimize an application is to characterize its most
+significant performance problem. First profile the application with
+`vmprofile` and then interpret the results using this step-by-step
+method:
+
+```lua
+-- RaptorJIT optimization procedure expressed as pseudo-code.
+function optimize ()
+   -- Spending nontrivial time in the interpreter? That's a bug - stop it!
+   if vmprofile.interpreted > 1% then
+      eliminate_interpreter()
+   -- Spend time in the garbage collector? That's a bug too - stop it!
+   elseif vmprofile.garbagecollection > 1% then
+      eliminate_allocation()
+   -- If you are not stuck in interpreter or GC then take a deeper look...
+   else
+      -- Spend time in JIT code? Check how efficient it is.
+      if vmprofile.head + vmprofile.loop > moderate then 
+         optimize_jit()
+      end
+      -- Spend time in the FFI? Check how efficient that is.
+      if vmprofile.ffi > moderate then
+          optimize_ffi()
+      end
+   end
+end
+```
+
+The control flow is significant in this method. For example, if the
+application is spending time in the interpreter then _that_ is the
+problem. You would need to eliminate this interpreter use before it
+makes sense look into other issues like garbage collection (which is
+often a secondary effect of running interpreted.)
+
+## Eliminate interpreter mode
+
+If your application is spending significant time in the interpreter
+then you need to look for _blacklisted_ code.
+
+Code is blacklisted when repeated attempts at JIT compilation all fail.
```

Comment:
Thanks!

I have pushed an expanded version of the "Eliminate interpreter mode" section now that focuses on aborts rather than blacklists (since we can't depend on those being present.) How does it look to you?

I added a simple example based on `-jv` but I feel it is a little unrealistic for applications like Snabb due to the usual "needle in a haystack" problem. The more code you have, the more JIT logs you have, the harder it is to focus on what actually matters. So I am still thinking about how to make a high-level visualization that points directly to the relevant information e.g. "NYI on line X is the root cause of Y% of the interpreter time in sub-profile FOO of the application."